File: 26 SAH Building/finished/src/obj_mesh.ts
import { vec3, vec2 } from "gl-matrix";
import { Triangle } from "./triangle";
import { Node } from "./node";
import { aabb } from "./aabb";

export class Bin {
    bounds: aabb;
    primitiveCount: number = 0; 
};

export class ObjMesh {

    v: vec3[]
    vt: vec2[]
    vn: vec3[]

    triangles: Triangle[]
    color: vec3

    triangleIndices: number[]
    nodes: Node[]
    nodesUsed: number

    minCorner: vec3
    maxCorner: vec3

    constructor() {

        this.v = [];
        this.vt = [];
        this.vn = [];

        this.triangles = [];

        this.minCorner = [ 999999,  999999,  999999];
        this.maxCorner = [-999999, -999999, -999999];
    }

    async initialize(color: vec3, url: string) {

        this.color = color;
        await this.readFile(url);
        this.v = [];
        this.vt = [];
        this.vn = [];
        this.buildBVH();
    }

    async readFile(url: string) {

        var result: number[] = [];

        const response: Response = await fetch(url);
        const blob: Blob = await response.blob();
        const file_contents = (await blob.text())
        const lines = file_contents.split("\n");

        lines.forEach(
            (line) => {
                //console.log(line);
                if (line[0] == "v" && line[1] == " ") {
                    this.read_vertex_data(line);
                }
                else if (line[0] == "v" && line[1] == "t") {
                    this.read_texcoord_data(line);
                }
                else if (line[0] == "v" && line[1] == "n") {
                    this.read_normal_data(line);
                }
                else if (line[0] == "f") {
                    this.read_face_data(line, result);
                }
            }
        )
    }

    read_vertex_data(line: string) {

        const components = line.split(" ");
        // ["v", "x", "y", "z"]
        const new_vertex: vec3 = [
            Number(components[1]).valueOf(),
            Number(components[2]).valueOf(),
            Number(components[3]).valueOf()
        ];

        this.v.push(new_vertex);

        vec3.min(this.minCorner, this.minCorner, new_vertex);
        vec3.max(this.maxCorner, this.maxCorner, new_vertex);
    }

    read_texcoord_data(line: string) {

        const components = line.split(" ");
        // ["vt", "u", "v"]
        const new_texcoord: vec2 = [
            Number(components[1]).valueOf(),
            Number(components[2]).valueOf()
        ];

        this.vt.push(new_texcoord);
    }

    read_normal_data(line: string) {

        const components = line.split(" ");
        // ["vn", "nx", "ny", "nz"]
        const new_normal: vec3 = [
            Number(components[1]).valueOf(),
            Number(components[2]).valueOf(),
            Number(components[3]).valueOf()
        ];

        this.vn.push(new_normal);
    }

    read_face_data(line: string, result: number[]) {

        line = line.replace("\n", "");
        const vertex_descriptions = line.split(" ");
        // ["f", "v1", "v2", ...]
        /*
            triangle fan setup, eg.
            v1 v2 v3 v4 => (v1, v2, v3), (v1, v3, v4)

            no. of triangles = no. of vertices - 2
        */

       const triangle_count = vertex_descriptions.length - 3; // accounting also for "f"
       for (var i = 0; i < triangle_count; i++) {
            var tri: Triangle = new Triangle();
            this.read_corner(vertex_descriptions[1], tri);
            this.read_corner(vertex_descriptions[2 + i], tri);
            this.read_corner(vertex_descriptions[3 + i], tri);
            tri.color = this.color;
            tri.make_centroid();
            this.triangles.push(tri);
       }
    }

    read_corner(vertex_description: string, tri: Triangle) {
        const v_vt_vn = vertex_description.split("/");
        const v = this.v[Number(v_vt_vn[0]).valueOf() - 1];
        const vt = this.vt[Number(v_vt_vn[1]).valueOf() - 1];
        const vn = this.vn[Number(v_vt_vn[2]).valueOf() - 1];
        tri.corners.push(v);
        tri.normals.push(vn);
    }

    buildBVH() {

        this.triangleIndices = new Array(this.triangles.length)
        for (var i:number = 0; i < this.triangles.length; i += 1) {
            this.triangleIndices[i] = i;
        }

        this.nodes = new Array(2 * this.triangles.length - 1);
        for (var i:number = 0; i < 2 * this.triangles.length - 1; i += 1) {
            this.nodes[i] = new Node();
        }

        var root: Node = this.nodes[0];
        root.leftChild = 0;
        root.primitiveCount = this.triangles.length;
        this.nodesUsed = 1

        this.updateBounds(0);
        this.subdivide(0);
    }

    updateBounds(nodeIndex: number) {

        var node: Node = this.nodes[nodeIndex];
        node.minCorner = [999999, 999999, 999999];
        node.maxCorner = [-999999, -999999, -999999];

        for (var i: number = 0; i < node.primitiveCount; i += 1) {
            const triangle: Triangle = this.triangles[this.triangleIndices[node.leftChild + i]];

            triangle.corners.forEach(
                (corner: vec3) => {

                    vec3.min(node.minCorner, node.minCorner, corner);
                    vec3.max(node.maxCorner, node.maxCorner, corner);
                }
            )
        }
    }

    EvaluateSAH(nodeIndex: number, axis: number, candidatePos: number): number {
        var leftBox: aabb = new aabb();
        var rightBox: aabb = new aabb();
        var leftCount = 0, rightCount = 0;
        const node = this.nodes[nodeIndex];
        
        for (let i = 0; i < node.primitiveCount; i += 1) {
            var triangle = this.triangles[this.triangleIndices[node.leftChild + i]];

            if (triangle.centroid[axis] < candidatePos) {
                leftCount += 1;
                leftBox.grow(triangle.corners[0]);
                leftBox.grow(triangle.corners[1]);
                leftBox.grow(triangle.corners[2]);
            }
            else {
                rightCount += 1;
                rightBox.grow(triangle.corners[0]);
                rightBox.grow(triangle.corners[1]);
                rightBox.grow(triangle.corners[2]);
            }
        }
        const cost = leftCount * leftBox.area() + rightCount * rightBox.area();
        if (cost > 0) {
            return cost;
        }
        else {
            return 9999999999;
        }
    }

    subdivide(nodeIndex: number) {

        var node: Node = this.nodes[nodeIndex];

        if (node.primitiveCount < 2) {
            return;
        }

        var extent: vec3 = [0, 0, 0];
        vec3.subtract(extent, node.maxCorner, node.minCorner);
        // determine split axis using SAH
        var bestAxis = -1;
        var bestPos = 0, bestCost = 9999999999;
        for (let axis = 0; axis < 3; axis += 1 ) { 
            var boundsMin = node.minCorner[axis];
            var boundsMax = node.maxCorner[axis];
            if (boundsMin == boundsMax) {
                continue;
            }
            var bins: Bin[] = new Array<Bin>(16);
            var scale = 16 / (boundsMax - boundsMin);
            for (let i = 0; i < node.primitiveCount; i += 1) {
                var triangle = this.triangles[this.triangleIndices[i + node.leftChild]];
                const binIdx = Math.min(15, Math.floor((triangle.centroid[axis] - boundsMin) * scale)); 
                bins[binIdx].primitiveCount += 1;
                bins[binIdx].bounds.grow(triangle.corners[0]);
                bins[binIdx].bounds.grow(triangle.corners[1]);
                bins[binIdx].bounds.grow(triangle.corners[2]);
            }
    
                const candidatePos = boundsMin + i * scale;
                const cost = this.EvaluateSAH(nodeIndex, axis, candidatePos);
                
                if (cost < bestCost) {
                    bestPos = candidatePos;
                    bestAxis = axis;
                    bestCost = cost;
                }
            }
        }


        var e: vec3 = vec3.create();
        vec3.subtract(e, node.maxCorner, node.minCorner);
        const parentArea = e[0] * e[1] + e[1] * e[2] + e[0] * e[2];
        const parentCost = node.primitiveCount * parentArea;
        if (parentCost < bestCost) {
            return;
        }
        
        var axis = bestAxis;
        var splitPosition = bestPos;

        var i: number = node.leftChild;
        var j: number = i + node.primitiveCount - 1;

        while (i <= j) {
            if (this.triangles[this.triangleIndices[i]].centroid[axis] < splitPosition) {
                i += 1;
            }
            else {
                var temp: number = this.triangleIndices[i];
                this.triangleIndices[i] = this.triangleIndices[j];
                this.triangleIndices[j] = temp;
                j -= 1;
            }
        }

        var leftCount: number = i - node.leftChild;
        if (leftCount == 0 || leftCount == node.primitiveCount) {
            return;
        }

        const leftChildIndex: number = this.nodesUsed;
        this.nodesUsed += 1;
        const rightChildIndex: number = this.nodesUsed;
        this.nodesUsed += 1;

        this.nodes[leftChildIndex].leftChild = node.leftChild;
        this.nodes[leftChildIndex].primitiveCount = leftCount;

        this.nodes[rightChildIndex].leftChild = i;
        this.nodes[rightChildIndex].primitiveCount = node.primitiveCount - leftCount;

        node.leftChild = leftChildIndex;
        node.primitiveCount = 0;

        this.updateBounds(leftChildIndex);
        this.updateBounds(rightChildIndex);
        this.subdivide(leftChildIndex);
        this.subdivide(rightChildIndex);
    }
}

File: 24 Higher Raytracing Performance/start/model_parser/obj_mesh.py
from config import *
import triangle
import node

class ObjMesh:

    def __init__(self, input_filepath: str, output_filepath: str):

        self.v: list[np.ndarray] = []

        self.triangles: list[triangle.Triangle] = []

        self.triangleIndices: list[int] = []
        self.nodes: list[node.Node] = []
        self.nodesUsed: int = 0

        self.minCorner: np.ndarray = pyrr.vector3.create(
            x =  999999, y =  999999, z =  999999, dtype = np.float32
        )
        self.maxCorner: np.ndarray = pyrr.vector3.create(
            x = -999999, y = -999999, z = -999999, dtype = np.float32
        )

        self.read_file(input_filepath)
        self.build_bvh()
        self.write_file(output_filepath)

    def read_file(self, filepath: str):

        with open(filepath, "r") as file:

            lines: list[str] = file.readlines()

            for line in lines:
                if (line[0] == "v" and line[1] == " "):
                    self.read_vertex_data(line)
                elif (line[0] == "f"):
                    self.read_face_data(line)

    def read_vertex_data(self, line: str):

        components = line.split(" ")
        #["v", "x", "y", "z"]
        new_vertex: np.ndarray = pyrr.vector3.create(
            x = float(components[1]),
            y = float(components[2]),
            z = float(components[3]),
            dtype = np.float32
        )

        self.v.append(new_vertex)

        self.minCorner = np.minimum(self.minCorner, new_vertex, dtype = np.float32)
        self.maxCorner = np.maximum(self.maxCorner, new_vertex, dtype = np.float32)

    def read_face_data(self, line: str):

        line = line.replace("\n", "")
        vertex_descriptions: list[str] = line.split(" ")
        # ["f", "v1", "v2", ...]
        """
            triangle fan setup, eg.
            v1 v2 v3 v4 => (v1, v2, v3), (v1, v3, v4)

            no. of triangles = no. of vertices - 2
        """
        
        triangle_count: int = len(vertex_descriptions) - 3 #accounting also for "f"
        for i in range(triangle_count):
            tri: triangle.Triangle = triangle.Triangle()
            self.read_corner(vertex_descriptions[1], tri)
            self.read_corner(vertex_descriptions[2 + i], tri)
            self.read_corner(vertex_descriptions[3 + i], tri)
            tri.make_centroid()
            self.triangles.append(tri)

    def read_corner(self, vertex_description: str, tri: triangle.Triangle):
        v_vt_vn: list[str] = vertex_description.split("/")
        v: np.ndarray = self.v[int(v_vt_vn[0]) - 1]
        tri.corners.append(v)

    def build_bvh(self):

        self.triangleIndices = [i for i in range(len(self.triangles))]

        self.nodes = [node.Node() for _ in range(2 * len(self.triangles) - 1)]

        root: node.Node = self.nodes[0]
        root.leftChild = 0
        root.primitiveCount = len(self.triangles)
        self.nodesUsed = 1

        self.update_bounds(0)
        self.subdivide(0)

    def update_bounds(self, nodeIndex: int):

        _node: node.Node = self.nodes[nodeIndex]
        _node.minCorner = pyrr.vector3.create(
            x = 999999, y = 999999, z = 999999, dtype = np.float32
        )
        _node.maxCorner = pyrr.vector3.create(
            x = -999999, y = -999999, z = -999999, dtype = np.float32
        )

        for i in range(_node.primitiveCount):
            _triangle: triangle.Triangle = self.triangles[
                self.triangleIndices[_node.leftChild + i]
            ]

            for corner in _triangle.corners:
                _node.minCorner = np.minimum(_node.minCorner, corner, dtype = np.float32)
                _node.maxCorner = np.maximum(_node.maxCorner, corner, dtype = np.float32)

    def subdivide(self, nodeIndex: int):

        print(f"Splitting node {nodeIndex}")

        _node: node.Node = self.nodes[nodeIndex]

        if (_node.primitiveCount < 2):
            return

        extent: np.ndarray = _node.maxCorner - _node.minCorner
        axis: int = 0
        if (extent[1] > extent[axis]):
            axis = 1
        if (extent[2] > extent[axis]):
            axis = 2

        splitPosition: float = _node.minCorner[axis] + extent[axis] / 2

        i: int = _node.leftChild
        j: int = i + _node.primitiveCount - 1

        while (i <= j):
            if (self.triangles[self.triangleIndices[i]].centroid[axis] < splitPosition):
                i += 1
            else:
                temp: int = self.triangleIndices[i]
                self.triangleIndices[i] = self.triangleIndices[j]
                self.triangleIndices[j] = temp
                j -= 1

        leftCount: int = i - _node.leftChild
        if (leftCount == 0 or leftCount == _node.primitiveCount):
            return

        leftChildIndex: int = self.nodesUsed
        self.nodesUsed += 1
        rightChildIndex: int = self.nodesUsed
        self.nodesUsed += 1

        self.nodes[leftChildIndex].leftChild = _node.leftChild
        self.nodes[leftChildIndex].primitiveCount = leftCount

        self.nodes[rightChildIndex].leftChild = i
        self.nodes[rightChildIndex].primitiveCount = _node.primitiveCount - leftCount

        _node.leftChild = leftChildIndex
        _node.primitiveCount = 0

        self.update_bounds(leftChildIndex)
        self.update_bounds(rightChildIndex)
        self.subdivide(leftChildIndex)
        self.subdivide(rightChildIndex)
    
    def write_file(self, filepath: str):

        with open(filepath, "w") as file:

            #minCorner
            file.write(f"min {self.minCorner[0]} {self.minCorner[1]} {self.minCorner[2]}\n")

            #maxCorner
            file.write(f"max {self.maxCorner[0]} {self.maxCorner[1]} {self.maxCorner[2]}\n")

            #nodesUsed
            file.write(f"nodes {self.nodesUsed}\n")

            #nodes
            for i in range(self.nodesUsed):
                _node: node.Node = self.nodes[i]
                line: str = f"node {_node.minCorner[0]} {_node.minCorner[1]} {_node.minCorner[2]} "
                line += f"{_node.maxCorner[0]} {_node.maxCorner[1]} {_node.maxCorner[2]} "
                line += f"{_node.leftChild} {_node.primitiveCount}\n"
                file.write(line)
            
            #triangleIndices
            for index in self.triangleIndices:
                file.write(f"triIndex {index}\n")